File: frontend/src/lib/components/system-tasks.ts
import { CommandGenerator, DeviceType } from './types'

export interface SystemTask {
  id: string
  label: string
  description: string
  deviceTypes?: DeviceType[]
  command: CommandGenerator
  requiresTerminal?: boolean
  needsWriteableRoot?: boolean
}

export const systemTasks: SystemTask[] = [
  {
    id: 'disable-updates',
    label: 'Disable Auto-Updates',
    description: 'Prevent automatic software updates',
    command: () => [
      {
        script: 'systemctl stop update-engine.service',
        description: 'Stop update service',
      },
      {
        script: 'systemctl disable update-engine.service',
        description: 'Disable update service',
      },
    ],
    requiresTerminal: true,
    needsWriteableRoot: true,
  },
  {
    id: 'enable-updates',
    label: 'Enable Auto-Updates',
    description: 'Re-enable automatic software updates',
    command: () => [
      {
        script: 'systemctl enable update-engine.service',
        description: 'Enable update service',
      },
      {
        script: 'systemctl start update-engine.service',
        description: 'Start update service',
      },
    ],
    requiresTerminal: true,
    needsWriteableRoot: false,
  },
]

File: internal/device/system_tasks.go
package device

import "reManager/internal/component"

type SystemTask struct {
	ID                 string
	Label              string
	Description        string
	DeviceTypes        []component.DeviceType
	Command            func(ctx component.CommandContext) []component.CommandResult
	RequiresTerminal   bool
	NeedsWriteableRoot bool
}

var SystemTasks = []SystemTask{
	{
		ID:          "enable-updates",
		Label:       "Enable Auto-Updates",
		Description: "Re-enable automatic software updates",
		Command: func(ctx component.CommandContext) []component.CommandResult {
			return []component.CommandResult{
				{
					Script:      "systemctl enable update-engine.service",
					Description: "Enable update service",
				},
				{
					Script:      "systemctl start update-engine.service",
					Description: "Start update service",
				},
			}
		},
		RequiresTerminal:   true,
		NeedsWriteableRoot: true,
	},
	{
		ID:          "disable-updates",
		Label:       "Disable Auto-Updates",
		Description: "Prevent automatic software updates",
		Command: func(ctx component.CommandContext) []component.CommandResult {
			return []component.CommandResult{
				{
					Script:      "systemctl stop update-engine.service",
					Description: "Stop update service",
				},
				{
					Script:      "systemctl disable update-engine.service",
					Description: "Disable update service",
				},
			}
		},
		RequiresTerminal:   true,
		NeedsWriteableRoot: true,
	},
	{
		ID:          "restart-xochitl",
		Label:       "Restart reMarkable UI",
		Description: "Restart the Xochitl UI service",
		Command: func(ctx component.CommandContext) []component.CommandResult {
			return []component.CommandResult{
				{
					Script:      "systemctl restart xochitl",
					Description: "Restart Xochitl service",
				},
			}
		},
		RequiresTerminal:   true,
		NeedsWriteableRoot: false,
	},
}

func GetSystemTask(id string) *SystemTask {
	for i := range SystemTasks {
		if SystemTasks[i].ID == id {
			return &SystemTasks[i]
		}
	}
	return nil
}
